# 간단한 예측을 해서 캐글에 제출해보기!!👀

In [1]:
# pandas를 import하고 train data와 test data를 불러온다.
import pandas as pd
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

In [2]:
# train data와 test data의 길이
train.shape, test.shape

((42000, 785), (28000, 784))

- train data는 컬럼 785개, 로우 42000개
- test data는 컬럼 784개, 로우 28000개

In [3]:
train.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

- train data에는 label이 포함되어 있다.

In [4]:
# train data의 label컬럼을 분리한다.
y = train['label']
X = train.drop('label',axis = 1)

In [5]:
X.shape, y.shape

((42000, 784), (42000,))

In [6]:
# test data와 train data를 분리하기 위해 sklearn의 train_test_split를 import한다.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [7]:
# data가 잘 나눠졌는지
X_train

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
31725       0       0       0       0       0       0       0       0       0   
22274       0       0       0       0       0       0       0       0       0   
17994       0       0       0       0       0       0       0       0       0   
13484       0       0       0       0       0       0       0       0       0   
20668       0       0       0       0       0       0       0       0       0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
6265        0       0       0       0       0       0       0       0       0   
11284       0       0       0       0       0       0       0       0       0   
38158       0       0       0       0       0       0       0       0       0   
860         0       0       0       0       0       0       0       0       0   
15795       0       0       0       0       0       0       0       0       0   

       pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  \
31725       0  ...         0         0         0         0         0   
22274       0  ...         0         0         0         0         0   
17994       0  ...         0         0         0         0         0   
13484       0  ...         0         0         0         0         0   
20668       0  ...         0         0         0         0         0   
...       ...  ...       ...       ...       ...       ...       ...   
6265        0  ...         0         0         0         0         0   
11284       0  ...         0         0         0         0         0   
38158       0  ...         0         0         0         0         0   
860         0  ...         0         0         0         0         0   
15795       0  ...         0         0         0         0         0   

       pixel779  pixel780  pixel781  pixel782  pixel783  
31725         0         0         0         0         0  
22274         0         0         0         0         0  
17994         0         0         0         0         0  
13484         0         0         0         0         0  
20668         0         0         0         0         0  
...         ...       ...       ...       ...       ...  
6265          0         0         0         0         0  
11284         0         0         0         0         0  
38158         0         0         0         0         0  
860           0         0         0         0         0  
15795         0         0         0         0         0  

[31500 rows x 784 columns]

In [8]:
y_train

31725    7
22274    8
17994    5
13484    8
20668    4
        ..
6265     9
11284    9
38158    2
860      6
15795    0
Name: label, Length: 31500, dtype: int64

In [9]:
# 랜덤 포레스트로 예측하기
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

rf_clf = RandomForestClassifier(random_state=42)
rf_clf = rf_clf.fit(X_train, y_train)
rf_prediction = rf_clf.predict(X_test)
print(rf_prediction[0:5])
print(y_test[0:5])
print(accuracy_score(y_test, rf_prediction))

/usr/local/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


[8 1 9 9 8]
5457     8
38509    1
25536    9
31803    9
39863    8
Name: label, dtype: int64
0.9360952380952381


In [11]:
# k-NN 알고리즘으로 예측하기
from sklearn.neighbors import KNeighborsClassifier

knn_clf = KNeighborsClassifier()
knn_clf.fit(X_train, y_train.values.ravel())

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [12]:
# 예측 해보기
prediction = knn_clf.predict(X_test)

In [13]:
# 정확도
print(accuracy_score(y_test,prediction))

0.9666666666666667


In [14]:
# test셋으로 결과 예측
result=knn_clf.predict(test)

In [15]:
# kaggle에서 제공해준 sample_submission 파일을 불러온다.
submission = pd.read_csv('/Users/Dahye/Desktop/notebooks/ml/sample_submission.csv')

In [16]:
submission.head()

ImageId  Label
0        1      0
1        2      0
2        3      0
3        4      0
4        5      0

In [17]:
# test셋으로 예측한 결과를 레이블에 넣어서 제출을 해보자.
submission['Label'] = result

In [18]:
# result가 잘 들어갔는지 확인
submission.head()

ImageId  Label
0        1      2
1        2      0
2        3      9
3        4      4
4        5      3

In [19]:
# result가 들어간 submission 데이터프레임을 csv파일로 저장한다.
# index=False 옵션을 지정해야 앞에 불필요한 인덱스가 생기지 않는다.
submission.to_csv('submission_knn.csv', index=False)